lensless/unrolled_admm.py

```diff
@@ -0,0 +1,248 @@
+# #############################################################################
+# unrolled_admm.py
+# =================
+# Authors :
+# Yohann PERRON [yohann.perron@gmail.com]
+# #############################################################################
+
+import numpy as np
+from lensless.trainable_recon import TrainableReconstructionAlgorithm
+from lensless.admm import soft_thresh, finite_diff, finite_diff_adj, finite_diff_gram
+from scipy import fft
+
+try:
+    import torch
+
+    torch_available = True
+except ImportError:
+    torch_available = False
+
+
+class UnrolledADMM(TrainableReconstructionAlgorithm):
+    """
+    Object for applying ADMM (Alternating Direction Method of Multipliers) with
+    a non-negativity constraint and a total variation (TV) prior.
+
+    Paper about ADMM: https://web.stanford.edu/~boyd/papers/pdf/admm_distr_stats.pdf
+
+    Slides about ADMM: https://web.stanford.edu/class/ee364b/lectures/admm_slides.pdf
+
+    """
+
+    def __init__(
+        self,
+        psf,
+        dtype=None,
+        n_iter=5,
+        mu1=1e-6,
+        mu2=1e-5,
+        mu3=4e-5,
+        tau=0.0001,
+        psi=None,
+        psi_adj=None,
+        psi_gram=None,
+        **kwargs,
+    ):
+        """
+
+        Parameters
+        ----------
+        psf : :py:class:`~torch.Tensor` of shape (H, W, C)
+            Point spread function (PSF) that models forward propagation.
+        dtype : float32 or float64
+            Data type to use for optimization. Default is float32.
+        n_iter : int, optional
+            Number of iterations to unrolled, by default 5
+        mu1 : float
+            Initial step size for updating primal/dual variables.
+        mu2 : float
+            Initial step size for updating primal/dual variables.
+        mu3 : float
+            Initial step size for updating primal/dual variables.
+        tau : float
+            Initial weight for L1 norm of `psi` applied to the image estimate.
+        psi : :py:class:`function`, optional
+            Operator to map image to a space that the image is assumed to be
+            sparse in (hence L1 norm). Default is to use total variation (TV)
+            operator.
+        psi_adj : :py:class:`function`
+            Adjoint of `psi`.
+        psi_gram : :py:class:`function`
+            Function to compute gram of `psi`.
+        """
+
+        super(UnrolledADMM, self).__init__(
+            psf, n_iter=n_iter, dtype=dtype, pad=False, norm="backward", reset=False, **kwargs
```

Comment:
similar to `ADMM` can put `pad` and `norm` as constructor inputs and default to these values